sql/parser: IsUnresolvedPlaceholder checks for nested parens

nvb opened the pull request:
Fixes #7331.

We now look for nested parentheses when checking if an `Expr`
is an unresolved placeholder. This fixes type checking of queries
like:

```
SELECT a FROM d.t WHERE a > ($1);
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7440)

<!-- Reviewable:end -->

nvb committed:
sql/parser: IsUnresolvedPlaceholder checks for nested parens

Fixes #7331.

We now look for nested parentheses when checking if an `Expr`
is an unresolved placeholder. This fixes type-checking of queries
like:

```
SELECT a FROM d.t WHERE a > ($1);
```

mjibson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7440#-:-KKyy9DO_l28ojulMtQR:1417088721)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7440#-:-KKyy_NpOzj4Ns5hvkOU:1417088721)*
<!-- Sent from Reviewable.io -->

nvb committed:
sql/parser: Create new StripParens function

This function strips the parenthesis off of a given expression. This is
useful when an expression type assertion is needed.

nvanbenschoten commented:
Stripping parenthesis from around an expression is useful for more than just this issue, so I pulled it out into a new function (and used it where possible). PTAL

tamird commented:
Reviewed 2 of 5 files at r2, 6 of 6 files at r4, 5 of 5 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7440)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/parser/type_check.go, line 217 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7440#-KLK1TP3O3FrrUCsG9Mo:-KLK1TP3O3FrrUCsG9Mp:966457087) ([raw file](https://github.com/cockroachdb/cockroach/blob/01076a8947489c5f22417dcac08514500c540c09/sql/parser/type_check.go#L217)):*
> ```Go
> 	// is passed to the child of the cast. There are two exceptions, described below.
> 	desired = NoTypePreference
> 	switch t := expr.Expr.(type) {
> ```

Why the change from the type switch to these functions?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7440)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/parser/type_check.go, line 217 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7440#-KLK1TP3O3FrrUCsG9Mo:-KLK9c4hB5vYvX-7s4Qv:-1176527188) ([raw file](https://github.com/cockroachdb/cockroach/blob/01076a8947489c5f22417dcac08514500c540c09/sql/parser/type_check.go#L217)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> Why the change from the type switch to these functions?
</details>
A type switch isn't sufficient now that `isUnresolvedPlaceholder` can be true for more than just the `Placeholder` type.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7440)*
<!-- Sent from Reviewable.io -->

mjibson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7440#-:-KLKAYFeb1qpw4kUkukO:1417088721)*
<!-- Sent from Reviewable.io -->
